### PR TITLE
Clean up the creational patterns for the TelemetryClient and the Senders

### PR DESCRIPTION
**Pull request by jkwatson**
Primarily, this moves the default connection timeouts to the HttpPoster implementations themselves, and adds a helpful factory method on the TelemetryClient. 

In addition, this adds a example class which shows the various ways you can configure the TelemetryClient.

This will possibly be a breaking change for someone who picked up the 0.5.1 release, unfortunately, but I think it cleans up the usage and is worth it.

### CODE REVIEW

File: telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java [hunk 1]
```diff
@@ -67,36 +69,77 @@ public static SenderConfigurationBuilder builder(String defaultUrl, String baseP
...
+     *
+     * @return this builder.
+     */
     public SenderConfigurationBuilder auditLoggingEnabled(boolean auditLoggingEnabled) {
       this.auditLoggingEnabled = auditLoggingEnabled;
       return this;
     }
 
+    /**
+     * Configure a secondary User-Agent value to use when sending data.
```
Comment (jeffalder) [inline telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java:133]: Please add a "why" here. Why, as a consumer of this, would I want or need to?

File: telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java [hunk 2]
```diff
@@ -67,36 +69,77 @@ public static SenderConfigurationBuilder builder(String defaultUrl, String baseP
...
+     *
+     * @return this builder.
+     */
     public SenderConfigurationBuilder secondaryUserAgent(String secondaryUserAgent) {
       this.secondaryUserAgent = secondaryUserAgent;
       return this;
     }
 
+    /** Create the resulting SenderConfiguration. */
```
Comment (jeffalder) [inline telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java:142]:
```suggestion
```

Restatement

File: telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java [hunk 3]
```diff
@@ -56,6 +57,7 @@ public static SenderConfigurationBuilder builder(String defaultUrl, String baseP
     return new SenderConfigurationBuilder(defaultUrl, basePath);
   }
 
+  /** A builder for SenderConfiguration instances. */
```
[the remark this answers is not in the document]
Comment (jkwatson) [reply telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java:60]: That's fair. We had been trying to at least put something, but I agree with you. It's not useful if it just repeats the obvious. I'll do some scrubbing of this.

File: telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java (hunk 1 above)
Comment (jkwatson) [reply telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java:126]: will update. thanks for the suggestion

File: telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java [hunk 4]
```diff
@@ -67,36 +69,77 @@ public static SenderConfigurationBuilder builder(String defaultUrl, String baseP
...
     }
 
+    /**
+     * Configure the endpoint URL for data to be sent to, not including the path. If the path is
+     * provided, it will be replaced with the default path.
+     *
+     * @param endpointUri A {@link URI}, path will be ignored.
+     * @return this builder.
+     */
+    public SenderConfigurationBuilder endpoint(URI endpointUri) throws MalformedURLException {
```
[the remark this answers is not in the document]
Comment (jkwatson) [reply telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java:109]: I agree. This is kind-of a mess at the moment. I do like having both options, but naming is hard. 

How about 

`endpointWithPath(URL url);`
and
`endpoint(String scheme, String host, int port)`

?

[the remark this answers is not in the document]
Comment (jeffalder) [reply telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java:109]: I like it! 👍

[the remark this answers is not in the document]
Comment (jkwatson) [reply telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java]: ok, updated with that. Please take a look!

File: telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java (hunk 1 above)
Comment (jkwatson) [reply telemetry_core/src/main/java/com/newrelic/telemetry/SenderConfiguration.java]: done

**Pull request merged**